src/coordinates/uniform_spherical.hpp

```diff
@@ -0,0 +1,621 @@
+
+  KOKKOS_FORCEINLINE_FUNCTION
+  Real h32f(const int j) const { return sintht_f_(j); }
+
+  KOKKOS_FORCEINLINE_FUNCTION
+  Real dh32vd2(const int j) const { return costht_c_(j); }
+
+  KOKKOS_FORCEINLINE_FUNCTION
+  Real dh32fd2(const int j) const { return costht_f_(j); }
+
+  //What is this?
+  KOKKOS_FORCEINLINE_FUNCTION
+  void GetAcc(const int k, const int j, const int i, const Real &rp,
+	      const Real &cosphip, const Real &sinphip, const Real &zp,
+	      Real &acc1, Real &acc2, Real &acc3) const
+  {
+    const Real cosdphi = cosphi_c_(j)*cosphip - sinphi_c_(j)*sinphip; //cos(x3v(k)-phip)
+    const Real sindphi = sinphi_c_(j)*cosphip - cosphi_c_(j)*sinphip; //sin(x3v(k)-phip)
+    //Psi = - 1.0/sqrt(r0^2 + rp^2 - 2r0*rp*sin(tht)cos(phi-phip) + zp^2- 2*r0*cos(tht)*zp
+    acc1 = (r_c_(i) - rp*sintht_c_(j)*cosdphi - costht_c_(j)*zp); //dPsi/dr0
+    acc2 = (-rp*costht_c_(j)*cosdphi + sintht_c_(j)*zp); //dPsi/(r0*dtht)
+    acc3 = (rp*sindphi);  //dPsi/(r0*sintht*dphi)
+  }
+
+  //----------------------------------------
+  // Position to Cell index
+  //----------------------------------------
+  KOKKOS_INLINE_FUNCTION
+  void Xtoijk(const Real &r, const Real &theta, const Real &phi, int &i, int &j, int &k) const {
+    i = (nx_[0] != 1) ? static_cast<int>(
+        std::floor((r - xmin_[0]) / dx_[0])) +
+      istart_[0] : istart_[0];
+    j = (nx_[1] != 1) ? static_cast<int>(std::floor(
+          (theta - xmin_[1]) / dx_[1])) +
+      istart_[1]
+      : istart_[1];
+    k = (nx_[2] != 1)? static_cast<int>(std::floor(
+          (phi - xmin_[2]) / dx_[2])) +
+      istart_[2]
+      : istart_[2];
+  }
+
+  const std::array<Real, 3> &GetXmin() const { return xmin_; }
+  const std::array<int, 3> &GetStartIndex() const { return istart_; }
+  const char *Name() const { return name_; }
+
+ //private:
+  std::array<int, 3> istart_, nx_;
+  std::array<Real, 3> xmin_, dx_, area_;
+  constexpr static const char *name_ = "UniformCylindrical";
+
+  const std::array<Real, 3> &Dx_() const { return dx_; }
+
+  ParArrayND<Real> r_c_, theta_c_, x1s2_, coord_vol_i_, coord_area2_i_, costht_f_, sintht_f_, sintht_c_, costht_c_, coord_area1_j_, cosphi_c_, sinphi_c_;
+  void InitCachedArrays() {
+    int nx1_tot = nx_[0]+2*Globals::nghost+1;
+    r_c_ = ParArrayND<Real>("centroid(r)", nx1_tot-1);
+    x1s2_ = ParArrayND<Real>("area1(r)", nx1_tot-1);
+    coord_vol_i_ = ParArrayND<Real>("volume(r)", nx1_tot-1);
+    coord_area2_i_ = ParArrayND<Real>("area2(r)", nx1_tot-1);
+    parthenon::par_for(
+      parthenon::loop_pattern_flatrange_tag, "initializeArraySph_r", 
+      parthenon::DevExecSpace(), 0, nx1_tot-2,
+      KOKKOS_LAMBDA(const int &i) {
+  //for (int i = 0; i < nx1_tot-1; i++) {
+      Real rm = xmin_[0] + i * dx_[0];
+      Real rp = rm + dx_[0];
+      r_c_(i) = 0.75*(std::pow(rp, 4) - std::pow(rm, 4)) /
+  (std::pow(rp, 3) - std::pow(rm, 3));
+      x1s2_(i) = TWO_3RD*(std::pow(rp,3) - std::pow(rm,3))/(SQR(rp) - SQR(rm));
```

Comment:
@Yurlungur here's the two lines where the centroids and so called "face-averaged quantities" differ in spherical coordinates. I'm headed to a lunch meeting so I'll have think about why these two are different later

Reply:
Ah I see---the former is volume integrated, the latter surface-integrated.

I guess then the question is philosophical/API: If I ask for X1 "centroid" value with topological element "face2" do I ever want the volume-averaged version? Or do I always want the face-averaged version? I *think* the answer is we always want the face-averaged version?

Reply:
I like that logic, the face position should probably be the face-averaged position.

The other place I can think of where face positions are relevant in AthenaPK is when we define a vector potential on faces then take a centered finite difference curl to get the magnetic field. We'd like the face position to correspond to a divergenceless curl and match the magnetic field we'd like to set as close as possible. I'm not sure if that would mean face-area averaged or cell-volume-averaged for the proper position on the face.

Reply:
I think for conservative prolongation on volumes, faces, and edges, you always want the centroid coordinates of the topological element itself (not the volume average for faces, for instance). At least at linear order choosing to model your function in the cell as $f(\vec{x}) = f_\textrm{centroid} + a_i (x^i - \bar x^i)$ with 
```latex
\bar x^i = \frac{\int_\text{element} d^nx \sqrt{\gamma} x^i}{\int_\text{element} d^nx \sqrt{\gamma}} 
```
where $\gamma_{ij}$ is the induced metric on the element, results in the volume averaged value of the prolongation function being equal to $f_\textrm{centroid}$. 

@Yurlungur: Why do you necessarily want to reconstruct to the face coordinate centers instead of the face centroids?

Reply:
Maybe I just wasn't thinking it through. Your argument makes sense to me @lroberts36

Reply:
@lroberts36 So if I understand correctly, we'd likely always want the position on the face to be the centroid of the face-area, not a centroid based on cell-volume. That's consistent with how Athena++ usually treats face positions although I'm suspicious that they've mixed them up in some place.

@Yurlungur Since we're cell centered the magnetic fields are currently cell-centered. That won't be the case once a CT-MHD scheme makes it into AthenaPK

Reply:
For prolongation, I think it is always true that we want the centroids of the topological element that we are prolongating on (at least for the default prolongation operator).

For reconstruction, we will have cell values and want to reconstruct them to faces (or for some upwind MHD algorithms have face values and want to reconstruct them on edges). The cell averaged value is probably most naturally interpreted as a point-wise value at the cell centroid. Then the reconstructed values on the face will be at two of the cell centroid coordinates and one of the face centroid coordinates, e.g. if I am reconstructing on the face normal to the $\theta$ direction the reconstructed position on the face will be $(r_\textrm{cell-centroid}, \theta_\textrm{face}, \phi_\textrm{cell-centroid})$. This maybe poses some issues for the order of the method, since for second order accuracy we would probably like to solve point-wise Riemann problems at the face centroids[^1], but maybe in practice this isn't a large source of error. 

All that being said, at least for Cartesian, cylindrical, and spherical coordinates, all of this information can easily be recovered from a single centroid coordinate template function 
```c++
Coordinates::Xc<CoordinateDirection dir, TopologicalElement el>(int k, int j, int i);
```
so that e.g. the reconstructed position of the lower face of a zone normal to the $\theta$ direction is given by `{coords.Xc<X1DIR, TE::CC>(k, j, i), coords.Xc<X2DIR, TE::F2>(k, j, i), coords.Xc<X3DIR, TE::CC>(k, j, i)}`.

[1:] I would be interested to know how/if Athena++ deals with this.